File: treinamentos/treinamento2/problema2.cpp
/**
 * II Treinamento para a Maratona de Programação 2016
 * UFT- Palmas
 *
 * Problema: Jogando Cartas Fora
 * Descrição: https://www.urionlinejudge.com.br/judge/pt/problems/view/1110
 *
 */

#include <deque>
#include <algorithm>
#include <vector>
#include <iostream>
 
using namespace std;

int main() {
    int m;

    while(scanf("%d", &m), m) { 
		deque<int> dq(m);
        iota(dq.begin(), dq.end(), 1);

        vector<int> v;    
		do {
            v.push_back(dq.front());
			dq.pop_front();
            int top = dq.front();
			dq.pop_front();
			dq.push_back(top);
		} while(dq.size() >= 2);

        cout << "Discarded cards: ";

        if(v.size() > 1) {
            cout << v[0];
            for(auto it = v.begin()+1; it != v.end(); ++it)
                cout << ", " << *it;
        }
		cout << "\nRemaining card: " << dq.front() << endl;
    }

    return 0;
}

File: treinamentos/treinamento2/Problema2.java
import java.io.*;
import java.util.*;

public class Problema2 {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		Integer m;
		
		while(true) {
			m = scan.nextInt();
			
			if(m == 0)
				break;
				
			Deque<Integer> dq = new LinkedList<Integer>();
			ArrayList<Integer> lst = new ArrayList<Integer>();
			
			for(Integer i = 1; i <= m; i++)
				dq.addLast(i);
			
			do {
				lst.add(dq.removeFirst());
				Integer top = dq.removeFirst();
				dq.addLast(top);
			} while(dq.size() >= 2);
			
			System.out.print("Discarded cards: ");
			 
			if(lst.size() > 1) {
	            System.out.printf("%d", lst.get(0));
	            
	            for(int i = 1; i < lst.size(); i++)
	            	System.out.printf(", %d", lst.get(i));
		    }
	
			System.out.printf("%nRemaining card: %d%n", dq.getFirst());
		}
	}
}
